#1828 Resolve Cypress Test Issues

srikant-ch5 opened the pull request:
Fixes #1828 

Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

nmgokhale commented:
@srikant-ch5 Any estimate on when this PR will be ready for review? After this PR is merged, we can create a `13.0.0 `production ready version.

srikant-ch5 commented:
> Hi @tomlyn Canvas Cypress tests are failing for below components because there is mismatch between position of element and expected position for example for command `cy.verifyNodeTransform("Binding (entry) node", 300, 349.5);` Error is `expected 500 to be close to 300 +/- 2` .
> 
> * canvas/decorators.cy.js
> * canvas/drag.cy.js
> * canvas/links.cy.js
> * canvas/nodes-resize.cy.js
> * canvas/snap-to-grid.cy.js
> * canvas/undo-redo.cy.js
> 
> @nmgokhale I am looking into properties/properties-tooltips.cy.js test as we discussed its due to tooltip not showing up in tables in UI.

Hi @tomlyn Could you please let me know what needs to be done here in above test cases is it okay to update those numbers ?

nmgokhale commented on canvas_modules/harness/cypress/e2e/canvas/clipboard.cy.js:
```diff
@@ -554,6 +554,8 @@ describe("Test clipboard with detachable links enabled", function() {
 		cy.rightClickToDisplayContextMenu(100, 400);
 		cy.clickOptionFromContextSubmenu("Edit", "Paste");
 
+		// eslint-disable-next-line cypress/no-unnecessary-waiting
```
Instead of adding this everywhere, can you add this config in https://github.com/elyra-ai/canvas/blob/main/canvas_modules/harness/cypress/.eslintrc.js

tomlyn commented:
Hi @srikant-ch5 I spent a lot of time looking at these issues in the afternoon and evening.

For decorators.cy.js -- this runs OK for me on my Mac except for this test: "Test positioning a decoration using distance on a straight line link" where I had to add cy.wait(10) before each verify command. I don't see the position problem with that file

For drag.cy.js -- I do see the position problem. I sent you a couple of videos in Slack. Unfortunately GitHub won't let me upload them here. 

For links.cy.js -- same this as with drag

I haven't looked at the other tests.

srikant-ch5 commented:
> Hi @srikant-ch5 I spent a lot of time looking at these issues in the afternoon and evening.
> 
> For decorators.cy.js -- this runs OK for me on my Mac except for this test: "Test positioning a decoration using distance on a straight line link" where I had to add cy.wait(10) before each verify command. I don't see the position problem with that file
> 
> For drag.cy.js -- I do see the position problem. I sent you a couple of videos in Slack. Unfortunately GitHub won't let me upload them here.
> 
> For links.cy.js -- same problem as with drag.cy.js
> 
> I haven't looked at the other tests.

Hi @tomlyn I have referred to canvas videos which you have shared to resolve below test cases and updated the PR.
canvas/decorators.cy.js
canvas/drag.cy.js
canvas/links.cy.js
canvas/nodes-resize.cy.js
canvas/snap-to-grid.cy.js

But I am still facing issues in canvas/undo-redo.cy.js .

@nmgokhale I have created a new issue https://github.ibm.com/NGP-TWC/wdp-abstract-canvas/issues/3755 related to tooltip. So currently only two test cases are failing properties-tooltips.cy.js and undo-redo.cy.js

nmgokhale commented:
@srikant-ch5 
> @nmgokhale I have created a new issue in abstract-canvas related to tooltip. So currently only two test cases are failing properties-tooltips.cy.js and undo-redo.cy.js

1. If the test in `properties-tooltips.cy.js` is failing because of tooltip issue, you can skip that test in this PR. And add a comment in abstract-canvas issue to enable that test after fixing the issue. 



2. Enable cypress tests in build using - Search for # TODO: Enable this while fixing issue - https://github.ibm.com/NGP-TWC/wdp-abstract-canvas/issues/3696 in the code to enable cypress tests.

tomlyn commented:
@srikant-ch5 I sent two files on Slack which should resolve the undoredo.cy.js problems.

srikant-ch5 commented:
> @srikant-ch5 I sent two files on Slack which should resolve the undoredo.cy.js problems.

Hi @tomlyn Thanks for the files. I have included those changes and now undo-redo test case is passing.
@nmgokhale I have skipped the failing tooltip test case. Now all the test cases are passing, could you please review this PR?

nmgokhale commented on canvas_modules/harness/cypress/e2e/canvas/nodes-resize.cy.js:
```diff
@@ -34,8 +34,14 @@ describe("Test adding resizing nodes", function() {
 	it("Test all attached and detached links appear OK after moving nodes around", function() {
 		cy.verifyNumberOfPortDataLinks(8);
 
+		// TODO -- Fix when autoselect is available.
```
There are 20 `TODO` comments in this PR. Remove TODO comments for all tests which are working.

srikant-ch5 replied:
Hi Neha I am keeping these TODO comments in the PR as we discussed.

tomlyn replied:
@nmgokhale This is the issue described here https://github.ibm.com/NGP-TWC/wdp-abstract-canvas/issues/3760
which we ought to fix if we can.  If we can fix it the tests would return to doing the automatic selection on drag.

tomlyn commented on canvas_modules/harness/cypress/e2e/canvas/drag.cy.js:
```diff
@@ -67,6 +67,8 @@ describe("Test to see if regular selection and drag behavior works " +
 
 		// Try dragging a node that is not selected -
 		// this should select the node being dragged and deselect the three selections
+		// TODO -- Fix when autoselect is available.
```
For each place where we added this comment please add another comment line underneath each one saying:
```
// See: https://github.ibm.com/NGP-TWC/wdp-abstract-canvas/issues/3760
```

tomlyn commented on canvas_modules/harness/cypress/e2e/canvas/links.cy.js:
```diff
@@ -697,7 +707,7 @@ describe("Test selectedLinkSelection = 'Detachable' configuration option", funct
 		// First create a new node on the canvas.
 		cy.clickToolbarPaletteOpen();
 		cy.clickCategory("Record Ops");
-		cy.dragNodeToPosition("Sample", 300, 450);
```
This doesn't look right. Looks like another instance where we need to click on the node before the drag.

tomlyn commented on canvas_modules/harness/cypress/e2e/canvas/node-insert-in-link.cy.js:
```diff
@@ -88,7 +88,7 @@ describe("Test dropping a node on a link with enableInsertNodeDroppedOnLink opti
 		// Now we do the same test expect we drag the node
 		// from within the canvas UI not palette
 		cy.verifyNumberOfPortDataLinks(4);
-		cy.dragNodeToPosition("Select", 490, 340);
```
This doesn't look right. Needs fixing

tomlyn commented on canvas_modules/harness/cypress/e2e/canvas/node-insert-in-link.cy.js:
```diff
@@ -100,7 +100,7 @@ describe("Test dropping a node on a link with enableInsertNodeDroppedOnLink opti
 		// This will be the node we move to the link.
 		cy.clickToolbarPaletteOpen();
 		cy.clickCategory("Record Ops");
-		cy.dragNodeToPosition("Sort", 500, 600);
```
This also doesn't not look right.

tomlyn commented on canvas_modules/harness/cypress/e2e/canvas/node-insert-in-link.cy.js:
```diff
@@ -200,6 +201,8 @@ describe("Test dropping a node, with existing links, on the canvas onto a link w
 		// Move the Sample node from the canvas onto the link between the
 		// Execution node and the Supernode.
+		// On the first Sort node moves and on the second move Sample node will move
+		cy.moveNodeToPosition("Sample", 400, 200);
```
This look like another instance where the initial selection on the node doesn't work.  Please try replacing this 'moveNodeToPosition with the code we've used else where where we get the node and click on it.

tomlyn commented on canvas_modules/harness/cypress/e2e/canvas/undo-redo.cy.js:
```diff
@@ -544,12 +556,19 @@ describe("Test undo/redo property values and title in common-properties", functi
 		//  Verification steps
-		verifySamplingRatioParameterValueInConsole("samplingRatio", null);
-		verifyErrorMessageForSamplingRatioParameterInConsole("error", "samplingRatio", "Select a sampling ratio");
+		// TODO -Cannoy berify an invalid entered value. See above.
```
Typo -- should be
TODO - Cannot verify an invalid entered value.

tomlyn commented on canvas_modules/harness/cypress/support/canvas/node-cmds.js:
```diff
@@ -212,6 +214,7 @@ Cypress.Commands.add("hoverOverNode", (nodeName) => {
 
 Cypress.Commands.add("hoverOverNodeLabel", (nodeName) => {
+	cy.wait(0);
 	cy.getNodeWithLabel(nodeName)
```
In other words, why is it necessary to wait for 0 seconds?

tomlyn reviewed (changes requested):
I did the review. As part of this change do you think it would be possible to upgrade to the latest version of Cypress, after we get the test running to our satisfaction?

srikant-ch5 commented:
> I did the review. As part of this change do you think it would be possible to upgrade to the latest version of Cypress, after we get the test running to our satisfaction?

Hi @tomlyn I have upgraded Cypress to v13.7.3 and all the test cases are passing in my local. Also I have updated code as per your feedback please review.
<img width="781" alt="Screenshot 2024-04-17 at 1 45 45 PM" src="https://github.com/elyra-ai/canvas/assets/16302714/ed65dbaf-346d-4825-84f4-f65c537f110a">

tomlyn commented on canvas_modules/harness/cypress/e2e/canvas/drag.cy.js:
```diff
@@ -80,6 +83,16 @@ describe("Test to see if regular selection and drag behavior works " +
+		cy.getCommentWithText(
+			"This canvas shows the 4 different node types and three link types: node links, " +
+			"association links and comments links."
+		).click();
```
Why is this clicking on the comment twice?

tomlyn commented on canvas_modules/harness/cypress/e2e/canvas/snap-to-grid.cy.js:
```diff
@@ -104,10 +116,13 @@ describe("Test resized comment snaps to grid", function() {
 		// during snap to grid may also move the position of the comment.
 		cy.setCanvasConfig({ "selectedSnapToGridType": "During" });
+		cy.wait(10);
+		cy.getCommentWithText("The 4 different node types").click();
```
This should have the comments added elsewhere for the same workaround.

tomlyn commented on canvas_modules/harness/cypress/e2e/canvas/undo-redo.cy.js:
```diff
@@ -55,6 +55,10 @@ describe("Test basic undo/redo operations", function() {
 		cy.clickNode("Select");
 		cy.clickToolbarAddComment();
+
+		// TODO - Fix when autoselect is available.
```
You missed the link to the issue here and for the change below.

tomlyn commented on canvas_modules/harness/cypress/e2e/canvas/undo-redo.cy.js:
```diff
@@ -114,6 +121,9 @@ describe("Test basic undo/redo operations", function() {
 		verifyRedoCommandExecuted("Var. File");
 
 		// Move comment on canvas
+		// TODO - Fix when autoselect is available.
```
Missed issue comment

tomlyn commented on canvas_modules/harness/cypress/e2e/canvas/undo-redo.cy.js:
```diff
@@ -544,12 +556,19 @@ describe("Test undo/redo property values and title in common-properties", functi
+		// TDOD - this part of the test is supposed to test an error in the sampling ration field
+		// However backspace no longer removes the default value from the field.
+		// Find a way to test an invlaid value in a field.
+		// cy.backspaceTextFieldValue("samplingRatio");
```
Please add a comment line here referencing the issue I opened for this problem which is: 
https://github.ibm.com/NGP-TWC/wdp-abstract-canvas/issues/3759

tomlyn reviewed (changes requested):
Hi Srikant - I competed my second review.

I noticed that there a lot of `cy.wait(10)` lines repeated many times after the same line of code. If we almost always have to add a wait after a function in the top level code, I think it would be better if the `cy.wait(10)` was added to the function being called rather than repeating it many times in the top level code. That way, the waits will not clutter up the code and also, if we write new testcases and reuse the function, we won't have to add in additional waits. Examples of what I mean are:
cy.clickOptionFromContextSubmenu
cy.clickColorFromContextSubmenu
cy.linkNodeOutputPortToNodeInputPort
You may be able to spot more ....

srikant-ch5 commented:
> Hi Srikant - I competed my second review.
> 
> I noticed that there a lot of `cy.wait(10)` lines repeated many times after the same line of code. If we almost always have to add a wait after a function in the top level code, I think it would be better if the `cy.wait(10)` was added to the function being called rather than repeating it many times in the top level code. That way, the waits will not clutter up the code and also, if we write new testcases and reuse the function, we won't have to add in additional waits. Examples of what I mean are: cy.clickOptionFromContextSubmenu cy.clickColorFromContextSubmenu cy.linkNodeOutputPortToNodeInputPort You may be able to spot more ....

Hi @tomlyn I have added cy.wait wherever test cases were failing but if we add wait in the commands then it would wait everytime its called which would slow down running time. For example there are 76 cy.wait's in total but if we add it in the command for verifyNumberOfLinks it would be called 98 times for one command and since there are many such commands it might take a bit longer to run cypress. Please let me know whether I should remove wait from top level and add those in commands. I have updated PR as per below feedback please review.

tomlyn commented:
Hi @srikant-ch5 That's why I said "If we **almost always have to add a wait** after a function in the top level code, " so if you are happy that we're not calling them unnecessarily from the top level code then that is OK with me.  BTW even adding 98 wait(10) would be 980 milliseconds. So less than a second on tests that usually take about 30mins to run.  

BTW - I don't see the green check mark against my name in the Reviewers list so I am assuming you'll click the little recycle button when you want me and Neha to review again. (Sorry if you already know this, I just want to make sure we are on the same page)
<img width="334" alt="image" src="https://github.com/elyra-ai/canvas/assets/15114439/94c2daad-bd4f-47d9-b4b1-3aaa7b82f555">

srikant-ch5 commented:
> Hi @srikant-ch5 That's why I said "If we **almost always have to add a wait** after a function in the top level code, " so if you are happy that we're not calling them unnecessarily from the top level code then that is OK with me. BTW even adding 98 wait(10) would be 980 milliseconds. So less than a second on tests that usually take about 30mins to run.
> 
> BTW - I don't see the green check mark against my name in the Reviewers list so I am assuming you'll click the little recycle button when you want me and Neha to review again. (Sorry if you already know this, I just want to make sure we are on the same page) <img alt="image" width="334" src="https://private-user-images.githubusercontent.com/15114439/323727114-94c2daad-bd4f-47d9-b4b1-3aaa7b82f555.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM1MDU4MjIsIm5iZiI6MTcxMzUwNTUyMiwicGF0aCI6Ii8xNTExNDQzOS8zMjM3MjcxMTQtOTRjMmRhYWQtYmQ0Zi00N2Q5LWI0YjEtM2FhYTdiODJmNTU1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE5VDA1NDUyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJiOTExNzliNGJlODkzZjNjMGQ1YjIwYTRjMmM3ZWM5ZDgxOTcyNDQ5ODg2MDNhZDc0N2QyNGRhN2RiYzYyYzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.etIurMotuz8zP0T724utAp26h0sVC-hynARW90UXOMo">

Hi @tomlyn we are calling wait only where its necessary so I hope this is fine for now. Please review once (I have raised review request again).
